tpm2/test/tpm2_test.go

```diff
@@ -1341,97 +1341,288 @@ func TestCreateAndCertifyCreationECC(t *testing.T) {
+	if err := nvDefine(rw); err != nil {
+		t.Fatalf("NVDefineSpace failed: %v", err)
+	}
+	defer nvUndefine(rw)
+	if err := nvWrite(rw); err != nil {
+		t.Fatalf("NVWrite failed: %v", err)
 	}
 
-	// Read all of the data with NVReadEx and compare to what was written
-	outdata, err := NVReadEx(rw, idx, HandleOwner, emptyPassword, 0)
+	data, err := nvRead(rw)
+	if err != nil {
+		t.Fatalf("NVRead failed: %v", err)
+	}
+	if !bytes.Equal(data, nvData) {
+		t.Errorf("data read from NV index does not match, got %x, want %x", nvData, data)
+	}
+}
+
+func TestNVReadEx(t *testing.T) {
+	rw := openTPM(t)
+	defer rw.Close()
+	if err := nvDefine(rw); err != nil {
+		t.Fatalf("NVDefineSpace failed: %v", err)
+	}
+	defer nvUndefine(rw)
+	if err := nvWrite(rw); err != nil {
+		t.Fatalf("NVWrite failed: %v", err)
+	}
+
+	data, err := NVReadEx(rw, nvIdx, HandleOwner, "", 0)
 	if err != nil {
 		t.Fatalf("NVReadEx failed: %v", err)
 	}
-	if !bytes.Equal(data, outdata) {
-		t.Fatalf("data read from NV index does not match, got %x, want %x", outdata, data)
+	if !bytes.Equal(data, nvData) {
+		t.Errorf("data read from NV index does not match, got %x, want %x", nvData, data)
 	}
+}
 
-	// Enable read lock
-	if err := NVReadLock(rw, HandleOwner, idx, emptyPassword); err != nil {
-		t.Fatalf("NVReadLock failed: %v (%T)", err, err)
+func TestNVReadLock(t *testing.T) {
+	rw := openTPM(t)
+	defer rw.Close()
+	if err := nvDefine(rw); err != nil {
+		t.Fatalf("NVDefineSpace failed: %v", err)
+	}
+	defer nvUndefine(rw)
+	if err := nvWrite(rw); err != nil {
+		t.Fatalf("NVWrite failed: %v", err)
 	}
 
-	// Read the data again. Should fail now because it's read-locked.
-	if _, err := NVReadEx(rw, idx, HandleOwner, emptyPassword, 0); err == nil {
-		t.Fatal("NVRead succeeded after NVReadLock")
+	if err := NVReadLock(rw, HandleOwner, nvIdx, emptyPassword); err != nil {
+		t.Fatalf("NVReadLock failed: %v (%T)", err, err)
+	}
+	// Can't compare raw fmt0 error code, because we decorate the error
+	// with cursor information in NVReadEx with fmt.Error.
+	// Instead, we look for the expected error string.
+	if _, err := nvRead(rw); err == nil {
+		t.Errorf("NVRead succeeded after NVReadLock")
 	} else if !strings.HasSuffix(err.Error(), ": NV access locked") {
-		t.Fatalf("NVRead: unexpected error; want RCNVLocked, got %v", err)
+		t.Errorf("NVRead: unexpected error; want RCNVLocked, got %v (%v)", err, reflect.TypeOf(err))
 	}
 }
 
-func TestPolicySecret(t *testing.T) {
+func TestNVWriteLock(t *testing.T) {
 	rw := openTPM(t)
 	defer rw.Close()
+	if err := nvDefine(rw); err != nil {
+		t.Fatalf("NVDefineSpace failed: %v", err)
+	}
+	defer nvUndefine(rw)
 
-	sessHandle, _, err := StartAuthSession(rw, HandleNull, HandleNull, make([]byte, 16), nil, SessionPolicy, AlgNull, AlgSHA256)
+	if err := NVWriteLock(rw, HandleOwner, nvIdx, emptyPassword); err != nil {
+		t.Fatalf("NVWriteLock failed: %v", err)
+	}
+	if err := nvWrite(rw); err == nil {
+		t.Errorf("NVRead succeeded after NVReadLock")
```

Comment:
Will fix this error string (should be NVWrite) along with any changes suggested by reviewers.